rl4co/models/common/constructive/nonautoregressive/encoder.py

```diff
@@ -0,0 +1,179 @@
+    def forward(self, graph: Batch) -> Tensor:  # type: ignore
+        # do not reuse the input value
+        edge_attr = graph.edge_attr  # type: ignore
+        for layer in self.linears:
+            edge_attr = self.act(layer(edge_attr))
+        graph.edge_attr = torch.sigmoid(self.output(edge_attr)) * 10  # type: ignore
+
+        heatmaps_logits = self._make_heatmaps(graph)
+        return heatmaps_logits
+
+    def _make_heatmaps(self, batch_graph: Batch) -> Tensor:  # type: ignore
+        graphs = batch_graph.to_data_list()
+        device = graphs[0].edge_attr.device
+        batch_size = len(graphs)
+        num_nodes = graphs[0].x.shape[0]
+
+        heatmaps_logits = torch.zeros(
+            (batch_size, num_nodes, num_nodes),
+            device=device,
+            dtype=graphs[0].edge_attr.dtype,
+        )
+
+        for index, graph in enumerate(graphs):
+            edge_index, edge_attr = graph.edge_index, graph.edge_attr
+            heatmaps_logits[index, edge_index[0], edge_index[1]] = edge_attr.flatten()
+
+        if self.undirected_graph:
+            heatmaps_logits = (heatmaps_logits + heatmaps_logits.transpose(1, 2)) * 0.5
+
+        return heatmaps_logits
+
+
+class NonAutoregressiveEncoder(ConstructiveEncoder):
+    """Anisotropic Graph Neural Network encoder with edge-gating mechanism as in Joshi et al. (2022).
+    This creates a heatmap # TODO
+
+        This model utilizes a multi-layer perceptron (MLP) approach to predict edge attributes directly from the input graph features,
+    which are then transformed into a heatmap representation to facilitate the decoding of the solution. The decoding process
+    is managed by a specified strategy which could vary from simple greedy selection to more complex sampling methods.
+
+    # TODO
+    Tip:
+        This decoder's performance heavily relies on the ability of the MLP to capture the dependencies between different
+        parts of the solution without the iterative refinement provided by autoregressive models. It is particularly useful
+        in scenarios where the solution space can be effectively explored in a parallelized manner or when the solution components
+        are largely independent.
+
+    Args:
+        env_name: Name of the environment used to initialize embeddings
+        embed_dim: Dimension of the node embeddings
+        num_layers: Number of layers in the encoder
+        init_embedding: Model to use for the initial embedding. If None, use the default embedding for the environment
+        edge_embedding: Model to use for the edge embedding. If None, use the default embedding for the environment
+        act_fn: The activation function to use in each GNNLayer, see https://pytorch.org/docs/stable/nn.functional.html#non-linear-activation-functions for available options. Defaults to 'silu'.
+        agg_fn: The aggregation function to use in each GNNLayer for pooling features. Options: 'add', 'mean', 'max'. Defaults to 'mean'.
+    """
```

Comment:
Agree, working on it! Should be ready in 1-2 hours ;)

Reply:
Done! You may check the latest~
Also, I separated the NAR GNN into a separate part of the zoo, since it can be a template for other models too. Open to better suggestions

Reply:
LGTM!